soco/discovery.py

```diff
@@ -239,3 +242,125 @@ def by_name(name):
         if device.player_name == name:
             return device
     return None
+
+
+def scan_network(max_threads=256, timeout=3.0, include_invisible=False):
+    """Scan all attached networks for Sonos devices.
+
+    Scans the IPv4 network attached to each interface to check for network devices
+    with port 1400 open. Check IPs in parallel threads for efficiency. Once the first
+    Sonos device is found, stop checking and use that device to find the other
+    devices. Returns a set of `SoCo` instances, or `None` if no Sonos devices are
+    discovered.
+
+    This function is intended for use when the usual discovery functions aren't
+    working, perhaps due to multicast problems on the network to which the SoCo
+    host is attached.
+
+    Args:
+        max_threads (int): The maximum number of threads to use when scanning the
+            network.
+        timeout (float): The network timeout in seconds to use when checking
+            each IP address for a Sonos device
+        include_invisible (bool): Whether to include invisible Sonos devices in
+            the set of devices returned.
+
+    Returns:
+        set: a set of `SoCo` instances, one for each zone found, or else `None`.
+    """
+
+    def is_ipv4_address(ip_address):
+        """Helper function to test for an IPv4 address."""
+        try:
+            ipaddress.IPv4Network(ip_address)
+            return True
+        except ValueError:
+            return False
+
+    def find_ipv4_networks():
+        """Helper function to return a list of unique IPv4 networks to which
+        this node is attached.
+        """
+        ipv4_net_list = []
+        adapters = ifaddr.get_adapters()
+        for adapter in adapters:
+            for ip_address in adapter.ips:
+                if is_ipv4_address(ip_address.ip):
+                    # Restrict to common domestic private IP ranges and sensible
+                    # netmasks. Experimental ... assumptions need to be tested.
+                    if (
```

Comment:
I'm wondering. Is this really necessary. On my system, I get only two ip addresses with network_prefix >= 24, which are localhost and the real local network. I'm just wondering because unless there is a reason to restrict it I don't think we should. SoCo might also be used e.g. for small office networks, which might have very different networks setups.

I think I would suggest to change it to just the network_prefix check AND that the ip address is not 127.0.0.1. On top of that add the option to disregard networks as a keyword argument for cases where it is a major performance hassle.

Reply:
I've changed the logic of this based on @rytilahti's comment below. I've also provided an optional `min_netmask` kwarg to constrain the network search space.